lib/seahorn/PredicateAbstraction.cc

```diff
@@ -63,22 +66,72 @@ namespace seahorn
 		boost::tribool result = fp.query ();
 //
 		if (result) outs () << "sat";
-		else if (!result) outs () << "unsat";
+		else if (!result)
+		{
+			outs() << "unsat\n";
+			HornDbModel absModel(new_db, fp);
+			converter.convert(absModel, oldModel, newToOldPredMap);
+			for(ExprMap::iterator it = oldModel.getRelToSolutionMap().begin(); it!=oldModel.getRelToSolutionMap().end(); ++it)
+			{
+				outs() << *(bind::fname(it->first)) << ": " << *(it->second) << "\n";
+			}
+		}
 		else outs () << "unknown";
 		outs () << "\n";
-//
-		printInvars (M);
 
 		return false;
 	}
 
+	bool HornModelConverter::convert (HornDbModel &in, HornDbModel &out, std::map<Expr, Expr> &newToOldPredMap)
+	{
+		ExprMap inRelToSolMap = in.getRelToSolutionMap();
```

Comment:
You should be able to get the list of relations from the database. I think there is already an api for that.
You can then get definition from an abstract model, construct the corresponding definition for the concrete relation and enter it into concrete model.

You definitely do not want anything that is specific to type of relation or arguments. `HornDbModel` should provide the same API as `getDeltaCover` since it is a generic representation of what we already have in the fixedpoint engine.